src/liballoc/sync.rs

```diff
@@ -328,6 +328,40 @@ impl<T> Arc<T> {
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+    /// to upgrade the weak reference before this function returns will result
+    /// in a `None` value.
+    #[inline]
+    #[unstable(feature = "arc_new_with", issue = "none")]
+    pub fn new_with(data_fn: impl FnOnce(&Weak<T>) -> T) -> Arc<T> {
+        // Start the weak pointer count at 2:
+        // - One for the pointer we'll pass into the constructor.
+        // - One for ourselves to ensure the memory is not freed.
+        let uninit_ptr: NonNull<_> = Box::leak(box ArcInner {
+            strong: atomic::AtomicUsize::new(0),
+            weak: atomic::AtomicUsize::new(1),
+            data: mem::MaybeUninit::<T>::uninit(),
+        })
+        .into();
+        let init_ptr: NonNull<ArcInner<T>> = uninit_ptr.cast();
+
+        let weak = Weak { ptr: init_ptr };
+        let data = data_fn(&weak);
+
+        unsafe {
+            let inner = init_ptr.as_ptr();
+            ptr::write(&raw mut (*inner).data, data);
+            (*inner).strong.fetch_add(1, Relaxed);
```

Comment:
We could assert (or at least debug_assert) that the old value is 0.

Also I wonder why `Relaxed` is sufficient. IMO `Relaxed` always needs a comment that explains why it is okay to use.

Reply:
I added a `debug_assert` - it doesn't seem like an overly likely thing to happen.

Re: ordering, you're correct: that should have been `Release`, and the "success" ordering parameter in `Weak::upgrade` needed to be updated to `Acquire`. I've added some justification in a comment. That change is one point against adding this functionality, although I don't expect it would make a signficant difference in practice.